File: cpp/samchon/protocol/InvokeParameter.cpp
#include <samchon/protocol/InvokeParameter.hpp>

#include <samchon/ByteArray.hpp>
#include <samchon/library/XML.hpp>
#include <samchon/library/Base64.hpp>
#include <samchon/library/StringUtil.hpp>

using namespace std;
using namespace samchon;
using namespace samchon::library;
using namespace samchon::protocol;

/* ----------------------------------------------------------
	CONSTRUCTORS
---------------------------------------------------------- */
InvokeParameter::InvokeParameter(const string &name, const char *pChar)
	: InvokeParameter(name, string(pChar))
{
}

//CUSTOM CONSTRUCTOR
InvokeParameter::InvokeParameter(const string &name, const string &type, const string &val)
	: super()
{
	this->name = name;
	this->type = type;
	this->str = val;
}

//MOVE CONSTRUCTORS
InvokeParameter::InvokeParameter(const string &name, string &&str)
	: super()
{
	this->name = name;
	this->type = "string";

	this->str = move(str);
}
InvokeParameter::InvokeParameter(const string &name, ByteArray &&byteArray)
	: super()
{
	this->name = name;
	this->type = "ByteArray";

	this->byteArray = move(byteArray);
}

/* ----------------------------------------------------------
	PROTECTED CONSTRUCTORS
---------------------------------------------------------- */
InvokeParameter::InvokeParameter()
	: super()
{
}

void InvokeParameter::construct(shared_ptr<XML> xml)
{
	if(xml->hasProperty("name") == true)
		this->name = xml->getProperty("name");
	else
		this->name = "";

	this->type = xml->getProperty("type");

	if(type == "XML")
		this->xml = xml->begin()->second->at(0);
	else if (type == "ByteArray")
	{
		size_t size = xml->getValue<size_t>();
		
		byteArray.reserve(size);
	}
	else
		this->str = xml->getValue();
}

auto InvokeParameter::reservedByteArraySize() const -> size_t
{
	return (size_t)stoull(StringUtil::between(this->str, "size: #"));
}
void InvokeParameter::setByteArray(ByteArray &&byteArray)
{
	this->byteArray = move(byteArray);
}

/* ----------------------------------------------------------
	GETTERS
---------------------------------------------------------- */
auto InvokeParameter::key() const -> string
{
	return name;
}

auto InvokeParameter::getName() const -> std::string
{
	return name;
}
auto InvokeParameter::getType() const -> std::string
{
	return type;
}

auto InvokeParameter::getValueAsXML() const -> shared_ptr<XML>
{
	return xml;
}

/* ----------------------------------------------------------
	EXPORTS
---------------------------------------------------------- */
auto InvokeParameter::TAG() const -> string
{
	return "parameter";
}

auto InvokeParameter::toXML() const -> shared_ptr<XML>
{
	shared_ptr<XML> &xml = super::toXML();

	if(name.empty() == false)
		xml->setProperty("name", name);
	xml->setProperty("type", type);

	if (type == "XML")
	{
		xml->push_back(this->xml);
	}
	else if (type == "ByteArray")
	{
		xml->setValue(byteArray.size());
	}
	else
	{
		xml->setValue(str);
	}

	return xml;
}
auto InvokeParameter::toSQL() const -> std::string
{
	//NAME, TYPE, VALUE
	std::string value;

	if (xml != nullptr)
		value = move(xml->toString());
	else
		value = this->str;

	std::string &sql = StringUtil::substituteSQL("({1}, {2}, {3})", name, type, value);
	return sql;
}

File: ts/src/samchon/protocol/InvokeParameter.ts
/// <reference path="../API.ts" />

/// <reference path="Entity.ts" />

namespace samchon.protocol
{
	/**
	 * A parameter belongs to an Invoke.
	 * 
	 * ![Class Diagram](http://samchon.github.io/framework/images/design/ts_class_diagram/protocol_message_protocol.png)
	 * 
	 * @author Jeongho Nam <http://samchon.org>
	 */
	export class InvokeParameter
		extends Entity
	{
		/**
		 * Name of the parameter.
		 *
		 * @details Optional property, can be omitted.
		 */
		protected name: string = "";

		/**
		 * Type of the parameter.
		 */
		protected type: string = "";

		/** 
		 * Value of the parameter.
		 */
		protected value: string | number | library.XML | Uint8Array = null;

		/**
		 * Default Constructor.
		 */
		public constructor();

		public constructor(val: number);
		public constructor(val: string);
		public constructor(val: library.XML);
		public constructor(val: Uint8Array);

		/**
		 * Construct from variable name and number value.
		 * 
		 * @param name
		 * @param val
		 */
		public constructor(name: string, val: number);
		public constructor(name: string, val: string);
		public constructor(name: string, val: library.XML);
		public constructor(name: string, val: Uint8Array);
		
		/* -------------------------------------------------------------------
			CONSTRUCTORS
		------------------------------------------------------------------- */
		public constructor(...args: any[])
		{
			super();

			// DEFAULT CONSTRUCTOR
			if (args.length == 0)
				return;

			// INITIALIZATION CONSTRUCTOR
			if (args.length == 1)
			{
				this.name = "";
				this.setValue(args[0]);
			}
			else
			{
				this.name = args[0];
				this.setValue(args[1]);
			}
		}

		/**
		 * @inheritdoc
		 */
		public construct(xml: library. XML): void
		{
			this.name = (xml.hasProperty("name")) ? xml.getProperty("name") : "";
			this.type = xml.getProperty("type");

			if (this.type == "XML")
				this.value = xml.begin().second.front();
			else if (this.type == "number")
				this.value = Number(xml.getValue());
			else if (this.type == "string")
				this.value = xml.getValue();
		}

		public setValue(value: number): void;
		public setValue(value: string): void;
		public setValue(value: library.XML): void;
		public setValue(value: Uint8Array): void;

		public setValue(value: number | string | library.XML | Uint8Array): void
		{
			this.value = value;

			if (value instanceof library.XML)
				this.type = "XML";
			else if (value instanceof Uint8Array)
				this.type = "ByteArray";
			else
				this.type = typeof value;
		}

		/* -------------------------------------------------------------------
			GETTERS
		------------------------------------------------------------------- */
		/**
		 * @inheritdoc
		 */
		public key(): any
		{
			return this.name;
		}

		/**
		 * Get name.
		 */
		public getName(): string
		{
			return this.name;
		}

		/**
		 * Get type.
		 */
		public getType(): string
		{
			return this.type;
		}

		/**
		 * Get value.
		 */
		public getValue(): any
		{
			return this.value;
		}

		/* -------------------------------------------------------------------
			EXPORTERS
		------------------------------------------------------------------- */
		/**
		 * @inheritdoc
		 */
		public TAG(): string
		{
			return "parameter";
		}
		
		/**
		 * @inheritdoc
		 */
		public toXML(): library.XML
		{
			let xml: library.XML = new library.XML();
			xml.setTag(this.TAG());

			if (this.name != "")
				xml.setProperty("name", this.name);
			xml.setProperty("type", this.type);

			// NOT CONSIDERED ABOUT THE BINARY DATA
			if (this.type == "XML")
				xml.push(this.value as library.XML);
			else if (this.type != "ByteArray")
				xml.setValue(this.value + "");

			return xml;
		}
	}
}